### ISSUE TITLE
Podman missinterprets %h symbol in the bind volume source when container created/started from a unit (BTRFS file system)

### ISSUE THREAD

**Issue by PavelSosin-320** [containers/podman#11547]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

When container that assumes bind volume mounting is created or run from the generated by podman generate systemd command the source path can be expressed by the absolute path or via % builtin symbols, %h for example. All % builtin "path" symbols like %t and %h are expanded into an absolute path, i.e. must be accepted by podman as a valid bind volume source. All subdirectories related to the %h, %v, etc. should be accepted. I hope, local driver and fuse.mount support this scenario. 

**Steps to reproduce the issue:**

1. Create a container from the docker.io/theiaide/theis image with bind volume as /home/user/project:/home/project
2. Generate Systemd unit from the container.
2. Put the unit into ~/.config/systemd/user and enable it
3. Create a "project workspace directory" ~/project
4. Modify unit .service file: add Condition=PathExists=%h/project (to be sure that bind volume source exists and its full path is correctly expanded) and the bind volume source expressed as -v %h/project:/home/project
5. Start the unit using systemctl start
**Describe the results you received:**
"Invalid reference" error message
**Describe the results you expected:**
It must work because systemd unit must be shareable between users

**Additional information you deem important (e.g. issue happens only occasionally):**
It works OK in WSL because the workspace is located in the host filesystem
**Output of `podman version`:**
Podman Version 3.3.1
API Version:  3.3.1
```

**Output of `podman info --debug`:**

```
(paste your output here)
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
(paste your output here)
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/master/troubleshooting.md)**


Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Fedora 4 Workstation on the laptop
security-opt lable=disable has been added
Standalone podman run  -v /home/pavelsosin/project:/home/project [TheiaIDE](https://hub.docker.com/r/theiaide/theia) is OK.

**Comment by PavelSosin-320** [comment]
Sorry! It may be unrelated because it happens regardless of the volume type. Simply, podman run and Podman container create return err code 125 when executed by rootless user, i.e. systemd unit on my Fedora WS in the user session, like in #https://github.com/containers/podman/issues/2197. Then, cid file is not created , ............. stop and rm don't work. It never happens in when podman runs in the terminal, only when from the unit using systemctl --user start, daemon-reexec, etc. Hint: Podman REST API service is running in the background.

**Comment by rhatdan** [comment]
@vrothberg PTAL

**Comment by PavelSosin-320** [comment]
@giuseppe I think that it is exactly the #https://github.com/containers/podman/issues/2172 because the main difference between working case onWSL distro i.e. ext4 filesystem and failing Fedora34 WS i.e. btrfs filesystem. But I expected that in 3.1 the configuration file should be corrected. Please, look at attachement.
[podman-info.log](https://github.com/containers/podman/files/7159217/podman-info.log)
I'm using Fedora WS GUI session, rootless user. Which conf file defines the overlay driver behavior in this case?

**Comment by giuseppe** [comment]
could you share the systemd service file generated by Podman?

**Comment by PavelSosin-320** [comment]
The recent working in rootless WSL version for Theia IDE with Podman
backend enabled, slightly modified to reference project workspace in the
user's home and starting Chrome in the kiosk mode.
Can be used as a test scenario.I use a pre-pooled Image ID to avoid pull
delay and .interaction with the user that is impossible inside systemd
units.
Everything works in WSL!!!

**Comment by PavelSosin-320** [comment]
The root cause is certainly related to the file system type: Fedora34 uses BTRFS by the default mounted at /home. If user data is located at /home too but not necessary on the btrfs subvolume the cross-fs mount will never work. The generated Unit checks that the Podman tree and container storage exist but it doesn't check or trigger mounting of subvolume or validate that user's data are in the same FS as container storage. The same problem appears in the anonymous volume scenario - nothing enforces that everything is on the same volume. storage.conf allows everything. WSL VM with its single ext4 filesystem works perfectly: no btrfs and Host folders are mounted using MS-specific mechanism. How bind mount should work on BTRFS and subvolumes???

**Comment by vrothberg** [comment]
> could you share the systemd service file generated by Podman?

@PavelSosin-320, please share the systemd service file.

**Comment by PavelSosin-320** [comment]
From the Podman on WSL instance with some comments and TODOs
[PodmanUnit.txt](https://github.com/containers/podman/files/7196471/PodmanUnit.txt)

**Comment by vrothberg** [comment]
Thanks, can you also the contents of run-r8df511c2cb034c33a1ec70d63b670529.service?

**Comment by PavelSosin-320** [comment]
Sorry for the long delay due to the holidays. I tried to run theia image via systemd-run as a transient unit and the result is:
[pavelsosin@Dell user]$ journalctl --unit run-u1276.service
-- Journal begins at Mon 2021-09-13 16:33:09 IDT, ends at Thu 2021-09-30 17:28:05 IDT. --
Sep 30 17:26:17 Dell systemd[1]: _Started /usr/bin/podman run -d -p 3000:3000 -v /home/pavelsosin/project:/home/project docker.io/theiaide/theia._
Sep 30 17:26:17 Dell podman[384240]: 2021-09-30 17:26:17.896394707 +0300 IDT m=+0.044832762 image pull  docker.io/theiaide/theia
Sep 30 17:26:18 Dell podman[384240]: 2021-09-30 17:26:18.338036042 +0300 IDT m=+0.486474147 container create ea48aca6642ccc7e87476b7611950d711efb28569efd8823a749e13a1d2f0c0e (image=docker.io/theiaide/theia:latest, name=determined_hertz)
Sep 30 17:26:19 Dell podman[384240]: 2021-09-30 17:26:19.592840034 +0300 IDT m=+1.741278171 container init ea48aca6642ccc7e87476b7611950d711efb28569efd8823a749e13a1d2f0c0e (image=docker.io/theiaide/theia:latest, name=determined_hertz)
Sep 30 17:26:19 Dell podman[384240]: 2021-09-30 17:26:19.937165559 +0300 IDT m=+2.085603734 container start ea48aca6642ccc7e87476b7611950d711efb28569efd8823a749e13a1d2f0c0e (image=docker.io/theiaide/theia:latest, name=determined_hertz)
Sep 30 17:26:19 Dell podman[384240]: ea48aca6642ccc7e87476b7611950d711efb28569efd8823a749e13a1d2f0c0e
Sep 30 17:26:20 Dell podman[384445]: 2021-09-30 17:26:20.229830119 +0300 IDT m=+0.072758624 container died ea48aca6642ccc7e87476b7611950d711efb28569efd8823a749e13a1d2f0c0e (image=docker.io/theiaide/theia:latest, name=determined_hertz)
Sep 30 17:26:22 Dell podman[384445]: 2021-09-30 17:26:22.572881081 +0300 IDT m=+2.415809589 container cleanup ea48aca6642ccc7e87476b7611950d711efb28569efd8823a749e13a1d2f0c0e (image=docker.io/theiaide/theia:latest, name=determined_hertz)
Sep 30 17:26:22 Dell systemd[1]: run-u1276.service: Deactivated successfully.
Sep 30 17:26:22 Dell systemd[1]: run-u1276.service: Consumed 1.134s CPU time
In other words, bind mount succeeded. This is good news. The bad news that possibly, either podman or storage driver search something in the unit environment that doesn't exist -"empty".
I also checked that no hidden dependency from systemd fs* target effects the bind because AssertMount /home is passed. BTRFS /home userspace FS is mounted, i.e. /home works as a BTRFS subvolume.

**Comment by PavelSosin-320** [comment]
Since all possible scenarios that I tested including Image volumes and anonymous volumes worked OK I suppose the root cause is that Podman parses -v option value exactly as described in the Documentation:
_Any source that does not begin with a . or / will be treated as the name of a named volume. If a volume with that name does not exist, it will be created .. in a blind way._ Systemd units placeholders that start from % don't conform with this pattern and are not substituted as Env variables. Unit, unlike systemd-run, ignores session environment - it runs without a shell. For example the equivalent of $(PWD) is WorkingDirectory=.

**Comment by vwbusguy** [comment]
How would it know to use btrfs driver vs overlay if storage.conf is ignored?

**Comment by PavelSosin-320** [comment]
Playing with RunC vs CRun I found that RunC has strong requirement that RunRoot where the bundle is stored must be on the TMPFS, i.e. /Run?User/ ... But Fedora ( and the future possible WSL Fedora distro based on WinBTRFS driver boxes users inside distro's /home FS that is always BTRFS. Does somebody know how to mount bind different FS type?
On my Fedora Desktop it is done via 3th FS - gvfs that was the real omnivore : it mounts  my cloud storage to my home area:
_gvfsd-fuse on /run/user/1000/gvfs type **fuse.gvfsd-fuse** (rw,nosuid,nodev,relatime,user_id=1000,group_id=1000)
ls -al /run/user/1000/gvfs_ 
_dr-x------.  1 pavelsosin pavelsosin   0 Jan  1  1970 'google-drive:host=gmail.com,user=pavel.sosin'_
Even it is hard to believe that the direct call for mount will succeed but I'm wondering why my /etc/fstab has no tmpfs entry that can be marked with users option and will allow to mount /run/user /  to the rootless home subvolume. I suppose that tmpfs is mounted by the certain systemd unit of type .mount running by the Systemd-user manager and available via PrivateTmp= directive, as it is described in the systemd docs. Otherwise, how rootless users have tmp dirs? 
Indeed, it hasn't ???
